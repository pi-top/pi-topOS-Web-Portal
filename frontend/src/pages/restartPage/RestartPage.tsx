import React, { useState, useEffect } from "react";
import { Line as ProgressBar } from "rc-progress";

import Layout from "../../components/layout/Layout";
import Spinner from "../../components/atoms/spinner/Spinner";

import rebootScreen from "../../assets/images/reboot-screen.png";
import styles from "./RestartPage.module.css";

import { runningOnWebRenderer } from "../../helpers/utils";
import ManualPowerOnDialogContainer from "./manualPowerOnDialog/ManualPowerOnDialogContainer";

export enum ErrorMessage {
  GlobalError = "Something went wrong while setting me up! Please click 'Restart' and contact support@pi-top.com if you experience any problems",
  RebootError = "I can't get to sleep! Please hold my power button down - that always makes me sleepy",
  TimeoutError = "It seems as though this is taking longer than expected. Try refreshing this webpage in your browser. Otherwise, try checking your connection and restarting your pi-top device."
}

export enum ServerStatusMessages {
  Waiting = "Rebooting. Please wait until this pi-top device is back online",
  Online = "The device is back online!"
}

export enum ExplanationMessages {
  OnWebRenderer = "Press 'restart' and I'll set some stuff up before rebooting. This might take up to a couple of minutes.",
  OnBrowser = "Press 'restart' to apply some final changes to your pi-top device and restart it.\n\nThis page will automatically update when the device is ready - this might take up to a couple of minutes, so don't go anywhere"
}

export type Props = {
  isWaitingForServer: boolean
  serverRebooted: boolean
  globalError: boolean;
  isSettingUpDevice: boolean;
  rebootError: boolean;
  progressPercentage: number;
  progressMessage: string;
  displayManualPowerOnDialog: boolean;
  onManualPowerOnDialogClose: () => void;
  setupDevice: () => void;
  onBackClick?: () => void;
};

export default ({
  isWaitingForServer,
  serverRebooted,
  globalError,
  isSettingUpDevice,
  rebootError,
  progressPercentage,
  progressMessage,
  displayManualPowerOnDialog,
  onManualPowerOnDialogClose,
  setupDevice,
  onBackClick,
}: Props) => {
  const [manualPowerOnDialogActive, setManualPowerOnDialogActive] = useState(false);

  useEffect(() => {
    setManualPowerOnDialogActive(displayManualPowerOnDialog);
  }, [displayManualPowerOnDialog])

  let errorMessage = "";
  if (globalError) {
    errorMessage = ErrorMessage.GlobalError;
  }

  if (rebootError) {
    errorMessage = isWaitingForServer ? ErrorMessage.TimeoutError : ErrorMessage.RebootError;
  }

  let explanationMessage = runningOnWebRenderer() ? ExplanationMessages.OnWebRenderer : ExplanationMessages.OnBrowser;

  return (
    <Layout
      banner={{
        src: rebootScreen,
        alt: "reboot-screen",
      }}
      prompt={
        <>
          Right, I need a quick <span className="green">nap</span>
        </>
      }
      explanation={explanationMessage}
      nextButton={{
        onClick: setupDevice,
        label: "Restart",
        disabled: isSettingUpDevice || rebootError || isWaitingForServer,
        hidden: isWaitingForServer || serverRebooted
      }}
      backButton={
        (globalError || !onBackClick)
          ? undefined
          : {
              onClick: onBackClick,
              disabled: isSettingUpDevice || isWaitingForServer,
              hidden: isWaitingForServer || serverRebooted
            }
      }
      className={styles.root}
    >

      <ManualPowerOnDialogContainer
        active={manualPowerOnDialogActive}
        onClose={() => {
          setManualPowerOnDialogActive(!manualPowerOnDialogActive);
          onManualPowerOnDialogClose();
        }}
      />

      {isSettingUpDevice && (
        <div className={styles.progress}>
          <ProgressBar
            percent={progressPercentage}
            strokeWidth={2}
            strokeColor="#71c0b4"
          />
          <span className={styles.message}>{progressMessage}</span>
        </div>
      )}
      {isWaitingForServer ? (
        <>
          <span className={styles.message}>{ServerStatusMessages.Waiting}</span>
          <Spinner size={40} />{" "}
        </>
      ):(
        serverRebooted && <span className={styles.message}>{ServerStatusMessages.Online}</span>
      )}

      {errorMessage && <span className={styles.error}>{errorMessage}</span>}
    </Layout>
  );
};
