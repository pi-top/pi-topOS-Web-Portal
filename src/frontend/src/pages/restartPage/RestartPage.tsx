import React from "react";
import { Line as ProgressBar } from "rc-progress";

import Layout from "../../components/layout/Layout";
import Spinner from "../../components/atoms/spinner/Spinner";

import rebootScreen from "../../assets/images/reboot-screen.png";
import styles from "./RestartPage.module.css";

export enum ErrorMessage {
  GlobalError = "Something went wrong while setting me up! Please click 'Restart' and contact support@pi-top.com if you experience any problems",
  RebootError = "I can't get to sleep! Please hold my power button down - that always makes me sleepy"
}

export enum ServerStatusMessages {
  Waiting = "Rebooting device, please wait until the unit is back online...",
  Online = "The device is back online!"
}

export type Props = {
  isWaitingForServer: boolean
  serverRebooted: boolean
  globalError: boolean;
  isSettingUpDevice: boolean;
  rebootError: boolean;
  progressPercentage: number;
  progressMessage: string;
  setupDevice: () => void;
  onBackClick?: () => void;
};

export default ({
  isWaitingForServer,
  serverRebooted,
  globalError,
  isSettingUpDevice,
  rebootError,
  progressPercentage,
  progressMessage,
  setupDevice,
  onBackClick,
}: Props) => {
  let errorMessage = "";
  if (globalError) {
    errorMessage = ErrorMessage.GlobalError;
  }

  if (rebootError) {
    errorMessage = ErrorMessage.RebootError;
  }

  return (
    <Layout
      banner={{
        src: rebootScreen,
        alt: "reboot-screen",
      }}
      prompt={
        <>
          Right, I need a quick <span className="green">nap</span>
        </>
      }
      explanation="Press restart and I'll set some stuff up before rebooting"
      nextButton={{
        onClick: setupDevice,
        label: "Restart",
        disabled: isSettingUpDevice || rebootError || isWaitingForServer,
        hidden: isWaitingForServer || serverRebooted
      }}
      backButton={
        (globalError || !onBackClick)
          ? undefined
          : {
              onClick: onBackClick,
              disabled: isSettingUpDevice || isWaitingForServer,
              hidden: isWaitingForServer || serverRebooted
            }
      }
      className={styles.root}
    >
      {isSettingUpDevice && (
        <div className={styles.progress}>
          <ProgressBar
            percent={progressPercentage}
            strokeWidth={2}
            strokeColor="#71c0b4"
          />
          <span className={styles.message}>{progressMessage}</span>
        </div>
      )}
      {isWaitingForServer ? (
        <>
          <span className={styles.message}>{ServerStatusMessages.Waiting}</span>
          <Spinner size={40} />{" "}
        </>
      ):(
        serverRebooted && <span className={styles.message}>{ServerStatusMessages.Online}</span>
      )}

      {errorMessage && <span className={styles.error}>{errorMessage}</span>}
    </Layout>
  );
};
