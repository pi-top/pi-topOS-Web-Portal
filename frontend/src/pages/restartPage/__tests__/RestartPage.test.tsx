import React from "react";
import {
  render,
  fireEvent,
  BoundFunction,
  GetByBoundAttribute,
  QueryByText,
  GetByText,
  RenderResult,
  QueryByBoundAttribute,
  wait,
} from "@testing-library/react";

import RestartPage, { Props, ErrorMessage, ExplanationMessages } from "../RestartPage";

describe("RestartPage", () => {
  let defaultProps: Props;
  let restartPage: HTMLElement;
  let getByAltText: BoundFunction<GetByBoundAttribute>;
  let queryByText: BoundFunction<QueryByText>;
  let getByText: BoundFunction<GetByText>;
  let queryByTestId: BoundFunction<QueryByBoundAttribute>;
  let rerender: RenderResult["rerender"];

  beforeEach(() => {
    defaultProps = {
      onBackClick: jest.fn(),
      setupDevice: jest.fn(),
      isSettingUpDevice: false,
      progressPercentage: 0.5,
      progressMessage: "I am setting up",
      rebootError: false,
      globalError: false,
      isWaitingForServer: false,
      serverRebooted: false,
      displayManualPowerOnDialog: false,
      piTopIpAddress: "pi-top.local",
      shouldDisplayConnectivityDialog: false,
      onManualPowerOnDialogClose: jest.fn(),
      shouldMoveAwayFromAp: false,
      checkingOnSameNetwork: false,
      onConnectivityDialogSkip: jest.fn(),
    };

    ({
      container: restartPage,
      getByAltText,
      queryByText,
      queryByTestId,
      getByText,
      rerender,
    } = render(<RestartPage {...defaultProps} />));
  });

  it("renders correct image", () => {
    expect(getByAltText("reboot-screen")).toMatchSnapshot();
  });

  it("renders prompt correctly", () => {
    const prompt = restartPage.querySelector(".prompt");
    expect(prompt).toMatchSnapshot();
  });

  it("renders explanation", () => {
    ExplanationMessages.OnBrowser.split('\n').map(function(item, _) {
      item && expect(getByText(item)).toBeInTheDocument();
    });
  });

  it("doesn't render error message", () => {
    expect(queryByTestId("error-message")).not.toBeInTheDocument();
  });

  it("doesn't render connectivity dialog", () => {
    expect(queryByTestId("connectivity-dialog")).not.toBeInTheDocument();
  });

  it("renders restart button", () => {
    expect(queryByText("Restart")).toBeInTheDocument();
  });

  it("renders back button", () => {
    expect(queryByText("Back")).toBeInTheDocument();
  });

  it("doesn't render skip button", () => {
    expect(queryByText("Skip")).not.toBeInTheDocument();
  });

  it("Back button isn't disabled", () => {
    expect(getByText("Back")).toHaveProperty("disabled", false);
  });

  it("Restart button isn't disabled", () => {
    expect(getByText("Restart").parentElement).toHaveProperty("disabled", false);
  });

  it("calls onBackClick on back button click", () => {
    fireEvent.click(getByText("Back"));

    expect(defaultProps.onBackClick).toHaveBeenCalled();
  });

  it('calls setupDevice on restart button click', () => {
    fireEvent.click(getByText('Restart'));

    expect(defaultProps.setupDevice).toHaveBeenCalled();
  })

  describe("when isSettingUpDevice is true and progress info is passed", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isSettingUpDevice: true,
      };

      rerender(<RestartPage {...defaultProps} />);
    });

    it("renders progress message", () => {
      expect(queryByText(defaultProps.progressMessage)).toBeInTheDocument();
    });

    it("renders progress bar correctly", () => {
      expect(restartPage.querySelector('.progress')).toMatchSnapshot();
    });

    it("disables the restart button", () => {
      expect(getByText("Restart").parentElement).toBeDisabled();
    });
  });

  describe("when rebootError is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        rebootError: true,
      };

      rerender(<RestartPage {...defaultProps} />);
    });

    it("renders correct error message", () => {
      expect(queryByText(ErrorMessage.RebootError)).toBeInTheDocument();
    });

    it("disables restart button", () => {
      expect(queryByText("Restart")?.parentElement).toBeDisabled();
    });
  });

  describe("when globalError is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        globalError: true,
      };

      rerender(<RestartPage {...defaultProps} />);
    });

    it("renders correct error message", () => {
      expect(queryByText(ErrorMessage.GlobalError)).toBeInTheDocument();
    });

    it("doesn't render skip button", () => {
      expect(queryByText("Skip")).not.toBeInTheDocument();
    });

    it("doesn't render back button", () => {
      expect(queryByText("Back")).not.toBeInTheDocument();
    })

    it("renders restart button", () => {
      expect(queryByText("Restart")).toBeInTheDocument();
    });

    it('calls setupDevice on restart button click', () => {
      fireEvent.click(getByText('Restart'));

      expect(defaultProps.setupDevice).toHaveBeenCalled();
    })
  });

  describe("when checkingOnSameNetwork is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        checkingOnSameNetwork: true,
      };

      rerender(<RestartPage {...defaultProps} />);
    });

    it("renders prompt correctly", () => {
      const prompt = restartPage.querySelector(".prompt");
      expect(prompt).toMatchSnapshot();
    });

    it("renders explanation", () => {
      expect(getByText(ExplanationMessages.CheckingConnectivity)).toBeInTheDocument();
    });

    it("restart button is disabled", () => {
      expect(queryByText("Restart")?.parentElement).toBeDisabled();
    });

    it("back button is hidden", () => {
      expect(queryByText("Back")).toHaveProperty("hidden");
    });
  });

  describe("when devices aren't on the same network", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        checkingOnSameNetwork: false,
        shouldMoveAwayFromAp: true,
        shouldDisplayConnectivityDialog: true,
      };

      rerender(<RestartPage {...defaultProps} />);
    });

    it("renders the dialog", () => {
      expect(queryByTestId("connectivity-dialog")).toBeInTheDocument();
    });

    it("displays the dialog", () => {
      expect(queryByTestId("connectivity-dialog")).not.toHaveClass("hidden");
    });

    it("disables restart button", () => {
      expect(queryByText("Restart")?.parentElement).toBeDisabled();
    });

    it("disables back button", () => {
      expect(queryByText("Back")).toBeDisabled();
    });
  });

  describe("when the only connection method is AP", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        checkingOnSameNetwork: false,
        shouldMoveAwayFromAp: false,
        shouldDisplayConnectivityDialog: true,
      };

      rerender(<RestartPage {...defaultProps} />);
    });

    it("renders the dialog", () => {
      expect(queryByTestId("connectivity-dialog")).toBeInTheDocument();
    });

    it("displays the dialog", () => {
      expect(queryByTestId("connectivity-dialog")).not.toHaveClass("hidden");
    });

    it("disables restart button", () => {
      expect(queryByText("Restart")?.parentElement).toBeDisabled();
    });

    it("disables back button", () => {
      expect(queryByText("Back")).toBeDisabled();
    });
  });

});
