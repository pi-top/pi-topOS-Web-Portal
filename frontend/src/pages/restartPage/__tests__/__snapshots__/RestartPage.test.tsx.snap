// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestartPage renders correct image 1`] = `
<img
  alt="reboot-screen"
  class="image"
  draggable="false"
  src="reboot-screen.png"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`RestartPage renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  Right, I need a quick 
  <span
    class="green"
  >
    nap
  </span>
</h1>
`;

exports[`RestartPage when checkingOnSameNetwork is true renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  Right, I need a quick 
  <span
    class="green"
  >
    nap
  </span>
</h1>
`;

exports[`RestartPage when isSettingUpDevice is true and progress info is passed renders progress bar correctly 1`] = `
<div
  class="progress"
>
  <svg
    class="rc-progress-line "
    preserveAspectRatio="none"
    viewBox="0 0 100 2"
  >
    <path
      class="rc-progress-line-trail"
      d="M 1,1
           L 99,1"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 1,1
           L 99,1"
      fill-opacity="0"
      stroke="#71c0b4"
      stroke-linecap="round"
      stroke-width="2"
      style="stroke-dasharray: 0.5px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear;"
    />
  </svg>
  <span
    class="message"
  >
    I am setting up
  </span>
</div>
`;
